vimeo/upload.py

```diff
@@ -20,82 +21,159 @@ class UploadVideoMixin(object):
+
+        https://tus.io/
+
+        Args:
+            attempt (:obj): requests object
+            path (string): path on disk to file
+            filename (string): name of the video file on vimeo.com
+
+        Returns:
+            string: The Vimeo Video URI of your uploaded video.
+
+        Raises:
+            VideoUploadFailure: If unknown errors occured when uploading your
+                video.
+        """
+        attempt = attempt.json()
+        upload_link = attempt.get('upload').get('upload_link')
 
-        return self._perform_upload(filename, ticket)
+        try:
+            with io.open(filename, 'rb') as fs:
+                tus_client = client.TusClient('https://files.tus.vimeo.com')
+                uploader = tus_client.uploader(file_stream=fs, url=upload_link)
+                uploader.upload()
+        except Exception as e:
+            raise exceptions.VideoUploadFailure(
+                e,
+                'Unexpected error when uploading through tus.'
+            )
+
+        return attempt.get('uri')
 
-    def _perform_upload(self, filename, ticket):
+    def __perform_upload(self, filename, ticket):
         """Take an upload ticket and perform the actual upload."""
         if ticket.status_code != 201:
             raise exceptions.UploadTicketCreationFailure(
-                ticket, "Failed to create an upload ticket")
+                ticket,
+                "Failed to create an upload ticket"
+            )
 
         ticket = ticket.json()
 
         # Perform the actual upload.
         target = ticket['upload_link']
-        free_quota = ticket['user']['upload_quota']['space']['free']
+        filesize = self.__check_upload_quota(filename)
         last_byte = 0
+
         # Try to get size of obj by path. If provided obj is not a file path
         # find the size of file-like object.
         try:
-            size = os.path.getsize(filename)
-            if size > free_quota:
-                raise exceptions.UploadQuotaExceeded(
-                    free_quota, 'Upload quota was exceeded.')
             with io.open(filename, 'rb') as f:
-                while last_byte < size:
+                while last_byte < filesize:
                     try:
-                        self._make_pass(target, f, size, last_byte)
+                        self.__make_pass(target, f, filesize, last_byte)
                     except requests.exceptions.Timeout:
                         # If there is a timeout here, we are okay with it,
                         # since we'll check and resume.
                         pass
-                    last_byte = self._get_progress(target, size)
+                    last_byte = self.__get_progress(target, filesize)
         except TypeError:
-            size = len(filename.read())
-            if size > free_quota:
-                raise exceptions.UploadQuotaExceeded(
-                    free_quota, 'Upload quota was exceeded.')
             f = filename
```

Comment:
I know this existed, before this PR. But I question whether this works. If you reach this point, the file stream has been closed and cleaned up since we exited the with statement. Copying the filename(string) to f then passing it to the "self.__make_pass" just seems odd. I imagine that the data sent on the put request would be the filename(string), unless the file is opened in the ```self.put``` call.

Reply:
I think this happens for if you pass a file pointer as `filename` instead of a path, which would throw a `TypeError` on `io.open(filename)` because that isn't a string.